python/plugins/processing/algs/qgis/TopoColors.py

```diff
@@ -0,0 +1,218 @@
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Nyall Dawson'
+__date__ = 'February 2017'
+__copyright__ = '(C) 2017, Nyall Dawson'
+
+# This will get replaced with a git SHA1 when you do a git archive323
+
+__revision__ = '$Format:%H$'
+
+import os
+import operator
+from collections import defaultdict, deque
+
+from qgis.core import (QgsField,
+                       QgsGeometry,
+                       QgsSpatialIndex,
+                       NULL)
+
+from qgis.PyQt.QtCore import (QVariant)
+
+from processing.core.GeoAlgorithm import GeoAlgorithm
+from processing.core.parameters import ParameterVector
+from processing.core.outputs import OutputVector
+from processing.tools import dataobjects, vector
+
+pluginPath = os.path.split(os.path.split(os.path.dirname(__file__))[0])[0]
+
+
+class TopoColor(GeoAlgorithm):
+    INPUT_LAYER = 'INPUT_LAYER'
+    OUTPUT_LAYER = 'OUTPUT_LAYER'
+
+    def defineCharacteristics(self):
+        self.name, self.i18n_name = self.trAlgorithm('Topological coloring')
+        self.group, self.i18n_group = self.trAlgorithm('Cartographic tools')
+        self.tags = self.tr('topocolor,colors,graph,adjacent,assign')
+
+        self.addParameter(ParameterVector(self.INPUT_LAYER,
+                                          self.tr('Input layer'), [dataobjects.TYPE_VECTOR_POLYGON]))
+
+        self.addOutput(OutputVector(self.OUTPUT_LAYER, self.tr('Colored'), datatype=[dataobjects.TYPE_VECTOR_POLYGON]))
+
+    def processAlgorithm(self, feedback):
+        layer = dataobjects.getObjectFromUri(
+            self.getParameterValue(self.INPUT_LAYER))
+
+        fields = layer.fields()
+        fields.append(QgsField('color_id', QVariant.Int))
+
+        writer = self.getOutputFromName(
+            self.OUTPUT_LAYER).getVectorWriter(
+            fields,
+            layer.wkbType(),
+            layer.crs())
+
+        # use a deque so we can drop features as we write them
+        # it's a bit friendlier on memory usage
+        features = deque(f for f in vector.features(layer))
+
+        topology, id_graph = self.compute_graph(features, feedback)
+        feature_colors = ColoringAlgorithm.balanced(topology, feedback)
+
+        max_colors = max(feature_colors.values())
+        feedback.pushInfo(self.tr('{} colors required').format(max_colors))
+
+        total = 20.0 / len(features)
```

Comment:
Good point - I should document that. It's basically just based on my tests with the algorithm and to try to balance the progress movement for the 3 stages.